engine/renderer/private/particles/particle_pipeline.cpp

```diff
@@ -87,92 +125,159 @@ void ParticlePipeline::RecordCommands(vk::CommandBuffer commandBuffer, uint32_t
+        commandBuffer.pushConstants(_pipelineLayouts[static_cast<uint32_t>(ShaderStages::eEmit)], vk::ShaderStageFlagBits::eCompute, 0, sizeof(EmitPushConstant), &_emitPushConstant);
         // +63 so we always dispatch at least once.
         commandBuffer.dispatch((_emitters[bufferOffset].count + 63) / 64, 1, 1);
     }
+    _emitters.clear();
 
+    vk::MemoryBarrier memoryBarrier {};
+    memoryBarrier.srcAccessMask = vk::AccessFlagBits::eShaderWrite;
+    memoryBarrier.dstAccessMask = vk::AccessFlagBits::eShaderRead;
     commandBuffer.pipelineBarrier(vk::PipelineStageFlagBits::eComputeShader, vk::PipelineStageFlagBits::eComputeShader, vk::DependencyFlags { 0 }, memoryBarrier, {}, {});
 
-    _emitters.clear();
-
     util::EndLabel(commandBuffer, vkContext->Dldi());
+}
+
+void ParticlePipeline::RecordSimulate(vk::CommandBuffer commandBuffer, float deltaTime)
+{
+    auto vkContext { _context->VulkanContext() };
 
-    // -- simulate shader pass --
     util::BeginLabel(commandBuffer, "Simulate particle pass", glm::vec3 { 255.0f, 105.0f, 180.0f } / 255.0f, vkContext->Dldi());
 
     commandBuffer.bindPipeline(vk::PipelineBindPoint::eCompute, _pipelines[static_cast<uint32_t>(ShaderStages::eSimulate)]);
 
     commandBuffer.bindDescriptorSets(vk::PipelineBindPoint::eCompute, _pipelineLayouts[static_cast<uint32_t>(ShaderStages::eSimulate)], 1, _particlesBuffersDescriptorSet, {});
     commandBuffer.bindDescriptorSets(vk::PipelineBindPoint::eCompute, _pipelineLayouts[static_cast<uint32_t>(ShaderStages::eSimulate)], 2, _instancesDescriptorSet, {});
 
-    _simulatePushConstant.deltaTime = deltaTime;
-    commandBuffer.pushConstants(_pipelineLayouts[static_cast<uint32_t>(ShaderStages::eSimulate)], vk::ShaderStageFlagBits::eCompute, 0, sizeof(float), &_simulatePushConstant);
+    _simulatePushConstant.deltaTime = deltaTime * 0.001f;
+    commandBuffer.pushConstants(_pipelineLayouts[static_cast<uint32_t>(ShaderStages::eSimulate)], vk::ShaderStageFlagBits::eCompute, 0, sizeof(SimulatePushConstant), &_simulatePushConstant);
 
     commandBuffer.dispatch(MAX_PARTICLES / 256, 1, 1);
 
+    vk::MemoryBarrier memoryBarrier {};
+    memoryBarrier.srcAccessMask = vk::AccessFlagBits::eShaderWrite;
+    memoryBarrier.dstAccessMask = vk::AccessFlagBits::eShaderRead;
     commandBuffer.pipelineBarrier(vk::PipelineStageFlagBits::eComputeShader, vk::PipelineStageFlagBits::eAllGraphics, vk::DependencyFlags { 0 }, memoryBarrier, {}, {});
 
     util::EndLabel(commandBuffer, vkContext->Dldi());
+}
+
+void ParticlePipeline::RecordRenderIndexed(vk::CommandBuffer commandBuffer, uint32_t currentFrame, const RenderSceneDescription& scene)
+{
+    auto vkContext { _context->VulkanContext() };
+    auto resources { _context->Resources() };
+    auto culledIndicesBuffer = resources->BufferResourceManager().Access(_culledInstancesBuffer);
+
+    // make sure the compute is done before the host reads from it
```

Comment:
Wait do we read data back from the device to the host? That sounds slow...

Reply:
Yeah at the moment we do for the drawIndexed to know how many instances we need. Assuming we'll transition to drawIndirect (which I would personally want to at some point) this readback should be eliminated with that